### PR TITLE
fields requirements changes

### PR DESCRIPTION
**Pull request by kevinkhao**
[IMP] Modifications related to changing "required" attribute on datamodels: includes SO generation, tests, components, demo data

Add tests for cases where data content is full/minimal/mixed
Fix state_code: search also according to country_id
Fix state_code, country_code validation and search

### CODE REVIEW

File: sale_import_base/components/importer_sale_channel.py [hunk 1]
```diff
@@ -50,17 +50,21 @@ def _validate_address(self, address, all_errors):
...
+                [("code", "=", address["country_code"])]
+            )
+            if len(country.ids) != 1:
+                errors += [_("Could not determine one country from country code")]
         if address.get("state_code"):
             state = self.env["res.country.state"].search(
-                [("code", "=", address["state_code"])]
+                [
+                    ("code", "=", address["state_code"]),
+                    ("country_id", "in", country and country.ids),
```
Comment (sebastienbeau) [inline sale_import_base/components/importer_sale_channel.py]: country variable can be not defined

File: sale_import_base/components/importer_sale_channel.py [hunk 2]
```diff
@@ -165,16 +169,19 @@ def _prepare_partner(self, data):
...
-            state = self.env["res.country.state"].search(
-                [("code", "=", data["state_code"])]
-            )
-            result["state_id"] = state.id
         if data.get("country_code"):
             country = self.env["res.country"].search(
                 [("code", "=", data["country_code"])]
             )
             result["country_id"] = country.id
+        if data.get("state_code"):
```
Comment (sebastienbeau) [inline sale_import_base/components/importer_sale_channel.py]: duplicated code

File: sale_import_base/datamodels/sale_order_amount.py [hunk 3]
```diff
@@ -7,6 +7,6 @@
 class SaleOrderAmountDatamodel(Datamodel):
     _name = "sale.order.amount"
 
-    amount_tax = fields.Decimal()
-    amount_untaxed = fields.Decimal()
-    amount_total = fields.Decimal()
+    amount_tax = fields.Decimal(required=True)
```
Comment (sebastienbeau) [inline sale_import_base/datamodels/sale_order_amount.py]: why?

Comment (sebastienbeau) [reply sale_import_base/datamodels/sale_order_amount.py]: should be not required

**Review by sebastienbeau — CHANGES_REQUESTED** [review]

File: sale_import_base/components/importer_sale_channel.py (hunk 1 above)
Comment (kevinkhao) [reply sale_import_base/components/importer_sale_channel.py]: fixed

File: sale_import_base/datamodels/sale_order_amount.py (hunk 3 above)
Comment (kevinkhao) [reply sale_import_base/datamodels/sale_order_amount.py]: sorry, commit in local was not yet pushed

File: sale_import_base/components/importer_sale_channel.py [hunk 4]
```diff
@@ -165,16 +107,27 @@ def _prepare_partner(self, data):
...
+            if len(country.ids) != 1:
+                raise ValidationError(
+                    _("Could not determine one country from country code")
+                )
             result["country_id"] = country.id
+        if data.get("state_code"):
+            state = self.env["res.country.state"].search(
+                [
+                    ("code", "=", data["state_code"]),
+                    ("country_id", "in", [result.get("country_id")]),
```
Comment (sebastienbeau) [inline sale_import_base/components/importer_sale_channel.py]: ("country_id", "=", result.get("country_id")),

Comment (kevinkhao) [reply sale_import_base/components/importer_sale_channel.py]: this whole part is made conditional on there being a country

**Pull request merged**